crates/wasi/src/cli.rs

```diff
@@ -0,0 +1,142 @@
+use std::rc::Rc;
+use std::sync::Arc;
+
+use wasmtime::component::ResourceTable;
+
+use crate::ResourceView;
+
+#[repr(transparent)]
+pub struct WasiCliImpl<T>(pub T);
+
+impl<T: WasiCliView> WasiCliView for &mut T {
+    type InputStream = T::InputStream;
+    type OutputStream = T::OutputStream;
+
+    fn cli(&mut self) -> &WasiCliCtx<T::InputStream, T::OutputStream> {
+        (**self).cli()
+    }
+}
+
+impl<T: WasiCliView> WasiCliView for Box<T> {
+    type InputStream = T::InputStream;
+    type OutputStream = T::OutputStream;
+
+    fn cli(&mut self) -> &WasiCliCtx<T::InputStream, T::OutputStream> {
+        (**self).cli()
+    }
+}
+
+impl<T: ResourceView> ResourceView for WasiCliImpl<T> {
+    fn table(&mut self) -> &mut ResourceTable {
+        self.0.table()
+    }
+}
+
+impl<T: WasiCliView> WasiCliView for WasiCliImpl<T> {
+    type InputStream = T::InputStream;
+    type OutputStream = T::OutputStream;
+
```

Comment:
I'm a bit wary of making the implementation more generic than it already is, so would it be possible to drop these associated types and hardwire the trait objects? Or are these needed to multiplex the p2/p3 implementation?

Reply:
Depends, if these per-proposal "View" traits are adopted by `p2` implementation, then the associated types would be used for exactly this purpose. Currently, only the p3 implementation uses these traits.

I'll move this trait to `wasmtime_wasi::p3::cli::WasiCliView` without the associated types for now and maybe we can revisit once we have all of WASI implemented in this repo.